#!/usr/bin/env python

"""
Carries out series of pair-wise comparisons between Bacillus subtilis 168
sequences from subtiwiki and those isolated from each genome, and also tblastn
comparison vs the genome sequence to ensure that the annotations are not
omitting any sequences.

Sequences are expected to be found in 'data/full/fasta/target_proteins',
generated by `extract_sequences.py`. BS168 sequences are available in the
'reference_sequences' directory 
"""

from subprocess import run
from pathlib import Path

from Bio import SeqIO
from Bio import SearchIO

import pandas as pd

reference_dir = Path('reference_sequences')
protein_dir   = Path('data/full/fasta/target_proteins')

def blast_genome(genome, gene, query, query_len):

    """
    Carries out blasts of protein sequence vs annotated protein sequence
    and genome to ensure annotations aren't missing/confusing sequences

    Required parameters:
        genome(pathlib::path): Path to genome sequence
        query(pathlib::Path): Path to query sequence
        gene(str): Gene name
        query_len(int): Length of query sequence # TODO can't this be parsed from teh outputs?
    
    Returns:
        hit_res(dict): Parsed hit details 
    """
    # TBLASTN of known protein vs genome to see if it is present
    # The hits will be variable for i.e. comP which are not well conserved

    command = ['tblastn','-query', query, '-subject', genome, '-outfmt', '5', '-out', '/tmp/result']
    run(command, check=True)

    bl_result = SearchIO.read('/tmp/result', 'blast-xml')
    if len(bl_result.hits):
        genome = bl_result.hits[0].id.split('|')[1]

        strand = '+'
        if '-' in str(bl_result.hits[0].hsps[0].hit_frame):
            strand = '-'

        hit_res = {
                'genome': genome,
                'description': bl_result.hits[0].description.split(' ',1)[1],
                'genome_query_len': query_len,
                'genome_align_len': bl_result.hits[0].hsps[0].aln_span,
                'genome_coverage': query_len/bl_result.hits[0].hsps[0].aln_span * 100,
                'genome_identity': bl_result.hits[0].hsps[0].ident_num,
                'genome_coords': f"{bl_result.hits[0].hsps[0].hit_start}-{bl_result.hits[0].hsps[0].hit_end}",
                'genome_strand': strand
            }
    else:
        hit_res = {'genome': genome}

    # BLASTP vs extracted sequence to make sure the annotation has found it correctly
    protein_file = protein_dir / gene / f"{genome}.fasta"
    if protein_file.exists():
        command = ['blastp', '-query', query, '-subject', protein_file, '-outfmt', '5', '-out', '/tmp/result']
        run(command, check = True)

        bl_result = SearchIO.read('/tmp/result', 'blast-xml')
        if len(bl_result.hits):

            hit_res['protein_align_len'] =  bl_result.hits[0].hsps[0].aln_span
            hit_res['protein_coverage'] =  query_len/bl_result.hits[0].hsps[0].aln_span * 100
            hit_res['protein_identity'] = bl_result.hits[0].hsps[0].ident_num

    # Flag cases where we seem to have better blast hits to genome than from identifiedo prediction
    hit_res['warning'] = False
    if ('protein_coverage' not in hit_res)  or \
        ('genome_coverage' not in hit_res) or \
        (hit_res['protein_coverage'] < hit_res['genome_coverage']) or \
        hit_res['genome_coverage'] < 90:
        hit_res['warning'] = True

    return(hit_res)

def main():

    gene_results = dict()

    for gene in ['comA', 'comP', 'comX', 'comQ', 'rapP']:
        results = list()

        query = reference_dir / Path(f'{gene}.fa')
        with open(query,'r', encoding='UTF-8') as fh:
            for record in SeqIO.parse(fh, format = 'fasta'):
                query_len = len(record.seq)
                break

        genomes = Path('data/full/fasta/genomes').glob('*fasta')
        for genome in genomes:
            results.append(blast_genome(genome, gene, query, query_len))

        results_df = pd.DataFrame(results)
        results_df = results_df.rename(columns = lambda x: f"{x}_{gene}")
        results_df = results_df.rename({f"genome_{gene}": "genome"}, axis=1)
        gene_results[gene] = results_df


    merged_df = pd.merge(gene_results['comA'], gene_results['comP'], on='genome', how = 'left')
    merged_df = pd.merge(merged_df, gene_results['comX'], on='genome', how = 'left')
    merged_df = pd.merge(merged_df, gene_results['comQ'], on='genome', how = 'left')
    merged_df = pd.merge(merged_df, gene_results['rapP'], on='genome', how = 'left')
    merged_df.drop(['description_comA','description_comP','description_comX','description_comQ','description_rapP'], 
        axis = 1, inplace = True)

    merged_df.to_csv('data/full/protein_coverage.txt', sep="\t", index=False)
    #with pd.ExcelWriter('data/full/protein_coverage.xlsx') as writer:
    #    merged_df.to_excel(writer, sheet_name = 'protein coverage', index = False)

if __name__ == "__main__":
    main()