from pathlib import Path

busco_lineage = 'bacillales_odb10'
cwd = Path('.').resolve()

GENOMES = Path('data/full/fasta').glob('*')
GENOMES = [genome.stem for genome in GENOMES]
BAKTA_DB_FILES = [ 'antifam.h3f', 'antifam.h3i', 'antifam.h3m', 'antifam.h3p', 'bakta.db',
        'expert-protein-sequences.dmnd', 'ncRNA-genes.i1f', 'ncRNA-genes.i1i', 'ncRNA-genes.i1m',
        'ncRNA-genes.i1p', 'ncRNA-regions.i1f', 'ncRNA-regions.i1i', 'ncRNA-regions.i1m',
        'ncRNA-regions.i1p', 'oric.fna', 'orit.fna', 'pfam.h3f', 'pfam.h3i', 'pfam.h3m', 'pfam.h3p',
        'rfam-go.tsv', 'rRNA.i1f', 'rRNA.i1i', 'rRNA.i1m', 'rRNA.i1p', 'sorf.dmnd'
    ]
BAKTA_DB = expand('databases/bakta/db/{file}', file = BAKTA_DB_FILES)
BAKTA_SUFFIXES = ['tsv', 'gff3', 'gbff', 'embl', 'fna', 'ffn', 'faa', 
    'hypotheticals.tsv', 'hypotheticals.faa', 'json', 'txt', 'png', 'svg']
BAKTA_ANNOTS = expand('data/full/annotations/{genome}/{genome}.{suffix}', genome = GENOMES, suffix = BAKTA_SUFFIXES)
BUSCO_DATA = expand('databases/busco/lineages/{busco_lineage}/dataset.cfg', busco_lineage = busco_lineage)
BUSCOS = expand('data/full/busco/{genome}/run_{lineage}/full_table.tsv', genome = GENOMES, lineage = busco_lineage)

rule all:
    input: BUSCOS

rule bakta_db:
    output: BAKTA_DB
    conda: 'envs/bakta.yaml'
    log: 'workflow/logs/bakta_db.log'
    shell: 'bakta_db download --type full -o databases/bakta > {log} 2>&1'

rule bakta:
    input: 
        fasta = 'data/full/fasta/{genome}.fasta', 
        db = BAKTA_DB
    output: 'data/full/annotations/{genome}/{genome}.tsv', 'data/full/annotations/{genome}/{genome}.gff3', 'data/full/annotations/{genome}/{genome}.gbff',
            'data/full/annotations/{genome}/{genome}.embl', 'data/full/annotations/{genome}/{genome}.fna', 'data/full/annotations/{genome}/{genome}.ffn',
            'data/full/annotations/{genome}/{genome}.faa', 'data/full/annotations/{genome}/{genome}.hypotheticals.tsv', 
            'data/full/annotations/{genome}/{genome}.hypotheticals.faa', 'data/full/annotations/{genome}/{genome}.json', 
            'data/full/annotations/{genome}/{genome}.txt', 'data/full/annotations/{genome}/{genome}.png', 'data/full/annotations/{genome}/{genome}.svg'
    conda: 'envs/bakta.yaml'
    log: 'workflow/logs/bakta.{genome}.log'
    shell: """
bakta --db databases/bakta/db \
      --prefix {wildcards.genome} \
      --genus Bacillus \
      --species subtilis \
      --compliant \
      --threads 8 \
      --force \
      --output data/full/annotations/{wildcards.genome} \
      {input.fasta} \
      > {log} 2>&1
    """

rule busco_lineage:
    output: BUSCO_DATA
    params: 
        lineage = busco_lineage
    conda: 'envs/busco.yaml'
    log: 'workflow/logs/busco_download.log'
    shell: 'busco --download_path databases/busco --download {params.lineage} > {log} 2>&1'

rule busco:
    input: 
        fasta = 'data/full/annotations/{genome}/{genome}.fna', 
        db = BUSCO_DATA
    output: 'data/full/busco/{genome}/run_bacillales_odb10/full_table.tsv'
    conda: 'envs/busco.yaml'
    params:
        lineage = busco_lineage,
        cwd = cwd
    log: 'workflow/logs/busco.{genome}.log'
    shell: """
busco -f -l {params.lineage} \
      --mode genome \
      -i {input.fasta} \
      -c 8 \
      -o data/full/busco/{wildcards.genome} \
      > {log} 2>&1
"""