from pathlib import Path
import pandas as pd
import re
from pprint import pprint

include: "rules/common.snmk"

busco_lineage = 'bacillales_odb10'
cwd = Path('.').resolve()
excel_files = list(cwd.glob('*.xlsx'))
genome_df = pd.read_excel(excel_files[0])
genome_df = genome_df[['accession', 'isolate']]

genome_df['isolate'] = genome_df.apply(lambda x: clean_strain(x['isolate']), axis = 1)
genome_info = genome_df.set_index('accession').to_dict()

GENOMES = Path('fasta').glob('*')
GENOMES = [genome.stem for genome in GENOMES]
BAKTA_DB = ['db/bakta.db']
BAKTA_SUFFIXES = ['tsv', 'gff3', 'gbff', 'embl', 'fna', 'ffn', 'faa', 
    'hypotheticals.tsv', 'hypotheticals.faa', 'json', 'txt', 'png', 'svg']
BAKTA_ANNOTS = expand('annotations/{genome}/{genome}.{suffix}', genome = GENOMES, suffix = BAKTA_SUFFIXES)
BUSCO_DATA = expand('busco_data/lineages/{busco_lineage}/dataset.cfg', busco_lineage = busco_lineage)
BUSCOS = expand('busco/{genome}/run_{lineage}/full_table.tsv', genome = GENOMES, lineage = busco_lineage)

rule all:
    input: BUSCOS

rule bakta_db:
    output: BAKTA_DB
    conda: 'envs/bakta.yaml'
    log: 'workflow/logs/bakta_db.log'
    shell: 'bakta_db download --type full > {log} 2>&1'

rule bakta:
    input: 'fasta/{genome}.fasta', BAKTA_DB
    output: 'annotations/{genome}/{genome}.tsv', 'annotations/{genome}/{genome}.gff3', 'annotations/{genome}/{genome}.gbff',
            'annotations/{genome}/{genome}.embl', 'annotations/{genome}/{genome}.fna', 'annotations/{genome}/{genome}.ffn',
            'annotations/{genome}/{genome}.faa', 'annotations/{genome}/{genome}.hypotheticals.tsv', 
            'annotations/{genome}/{genome}.hypotheticals.faa', 'annotations/{genome}/{genome}.json', 
            'annotations/{genome}/{genome}.txt', 'annotations/{genome}/{genome}.png', 'annotations/{genome}/{genome}.svg'
    conda: 'envs/bakta.yaml'
    log: 'workflow/logs/bakta.{genome}.log'
    shell: 'bin/bakta.sh -n {wildcards.genome} -g Bacillus -s subtilis -o annotations/{wildcards.genome} > {log} 2>&1'

rule busco_lineage:
    output: BUSCO_DATA
    params: 
        lineage = busco_lineage
    conda: 'envs/busco.yaml'
    log: 'workflow/logs/busco_download.log'
    shell: 'busco --download_path busco_data --download {params.lineage} > {log} 2>&1'

rule busco:
    input: 'annotations/{genome}/{genome}.fna', BUSCO_DATA
    output: 'busco/{genome}/run_bacillales_odb10/full_table.tsv'
    conda: 'envs/busco.yaml'
    params:
        lineage = busco_lineage,
        cwd = cwd
    log: 'workflow/logs/busco.{genome}.log'
    shell: 'bin/busco.sh -g {wildcards.genome} -l busco_data/lineages/{params.lineage} \
        -o busco/{wildcards.genome} > {log} 2>&1'